gateway/api/user/user.go

```diff
@@ -172,22 +205,47 @@ func Update(c *gin.Context) {
 	existingUser.Picture = req.Picture
 	existingUser.Status = string(req.Status)
 	existingUser.SlackID = req.SlackID
-	existingUser.Groups = req.Groups
 
-	if err := pgusers.New().Upsert(*existingUser); err != nil {
+	log.Debugf("updating user %s", userID)
+	if err := models.UpdateUser(existingUser); err != nil {
 		log.Errorf("failed updating user %s, err=%v", userID, err)
 		sentry.CaptureException(err)
 		c.JSON(http.StatusInternalServerError, gin.H{"message": err.Error()})
 		return
 	}
 
+	log.Debugf("updating user groups for user %s", userID)
+	if err := models.DeleteUserGroupsByUserID(existingUser.ID); err != nil {
+		log.Errorf("failed removing user groups for user %s, err=%v", userID, err)
+		sentry.CaptureException(err)
+		c.JSON(http.StatusInternalServerError, gin.H{"message": "failed removing user groups"})
+		return
+	}
+
+	if len(req.Groups) > 0 {
+		newUserGroups := []models.UserGroup{}
+		for i := range req.Groups {
+			newUserGroups = append(newUserGroups, models.UserGroup{
+				OrgID:  ctx.OrgID,
+				UserID: existingUser.ID,
+				Name:   req.Groups[i],
+			})
+		}
+		if err := models.InsertUserGroups(newUserGroups); err != nil {
+			log.Errorf("failed persisting user groups, err=%v", err)
+			sentry.CaptureException(err)
+			c.JSON(http.StatusInternalServerError, gin.H{"message": "failed persisting user groups"})
+			return
+		}
+	}
```

Comment:
This is the problem with not using transactions / full queries for cases likes this:
This method is now non idempotent and could leave the user in a inconsistent state.

We are trading "simplicity" over inconsistency.

Reply:
potential inconsistency, actually. And it's fine, like I've said in the other comment, the controller will pay the price. The one thing tho we might do to avoid this is to create a specific function at the model layer to handle a db transaction with all these operations. Still considering the overall availability of the DB using GORM I see this as a low-risk action to take. Whether doing it or not is a matter of detail